storage_test.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// TestZipAndUnzip validates that a directory can be compressed to a ZIP archive
+// using zipDir and subsequently decompressed using unzip to reproduce the
+// original directory structure and file contents.
+func TestZipAndUnZipDir(t *testing.T) {
+	// Create source directory with sample files.
+	sourceDir := filepath.Join(t.TempDir(), "test_corpus")
+	assert.NoError(t, os.Mkdir(sourceDir, 0o755))
+
+	fileContents := map[string][]byte{
+		"file1.txt": []byte("testing unzip"),
+		"file2.txt": []byte("testing zipDir"),
+	}
+	for name, data := range fileContents {
+		path := filepath.Join(sourceDir, name)
+		assert.NoError(t, os.WriteFile(path, data, 0o644))
+	}
+
+	// Initialize S3CorpusStore for zipping.
+	zipStore := &S3CorpusStore{
+		logger:    slog.New(slog.NewTextHandler(io.Discard, nil)),
+		corpusDir: sourceDir,
+	}
+
+	// Stream ZIP archive into a pipe.
+	pr, pw := io.Pipe()
+	go func() {
+		err := zipStore.zipDir(pw)
+		zipStore.logger.Error("zipDir error", "err", err)
+		pw.CloseWithError(err)
+	}()
+
+	// Write the archive to a separate temporary workspace.
+	archiveDir := t.TempDir()
+	zipPath := filepath.Join(archiveDir, "out.zip")
+
+	zipFile, err := os.Create(zipPath)
+	assert.NoError(t, err)
+
+	_, err = io.Copy(zipFile, pr)
+	assert.NoError(t, zipFile.Close())
+	assert.NoError(t, err)
```

Comment:
Nit: let's swap these lines for clarity

```suggestion
	assert.NoError(t, err)
	assert.NoError(t, zipFile.Close())
```